File: day-3/day3.py

with open("input") as f:
    data = [x.strip() for x in f.readlines()]

def count_trees(step_right, step_down):
    x = 0
    y = 0
    count = 0
    while y < len(data):
        count += data[y][x % len(data[0])] == '#'
        x += step_right
        y += step_down
    return count


print(count_trees(3, 1)) # part 1

print(count_trees(1, 1) * count_trees(3, 1) * count_trees(5, 1) * count_trees(7, 1) * count_trees(1, 2)) # part 2

File: day-3/day3.go
package main

import(
	"fmt"
	"os"
	"bufio"
	"strings"
	
)

func main(){
	file, _ := os.Open("input") 
    scanner := bufio.NewScanner(file) 
    scanner.Split(bufio.ScanLines) 

	var lines [][]string

    for scanner.Scan() {
		var s = strings.Split(scanner.Text(), "")   
		lines = append(lines, s)
	}

	fmt.Println(count_trees(lines, 3, 1)) // Part 1
	fmt.Println(count_trees(lines, 1, 1) * count_trees(lines, 3, 1) * count_trees(lines, 5, 1) * count_trees(lines, 7, 1) * count_trees(lines, 1, 2)) // Part 2
}

func count_trees(lines [][]string, right int, down int ) (int) {
	var x = 0 
	var y = 0
	var count = 0

	for y < len(lines) {
		if lines[y][x % len(lines[0])] == "#" {
			count++
		}
		x += right
		y += down
	}

	return count

}